### PR TITLE
Communicate Changes to UI

### PR DESCRIPTION
**Pull request by iamrakesh28**
Class CommunicateChanges is created

### CODE REVIEW

File: src/processing/CommunicateChange.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package processing;
+
+import java.util.*;
+import processing.IChanges;
+import processing.utility.*;
+
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java:5]: Add author's and reviewer's name

File: src/processing/CommunicateChange.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    
+	/*
+     * Computes the set of pixels which will have effect
+     * on the screen and notify the UI using the getChanges
+     * method of IChanges.
+     */
+	
+	// oldPixels -> previous pixels of the object
+	// newPixels -> new pixels of the object
+	public static void provideChanges(
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java:31]: multi line comments in /* * */ format.

File: src/processing/CommunicateChange.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			Identifier identifier, 
+			ArrayList<Pixel> oldPixels, 
+			ArrayList<Pixel> newPixels) {
+
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java:36]: multiple input parameters in separate lines? Check once with the Google code conventions.

File: src/processing/CommunicateChange.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+package processing;
+
+import java.util.*;
+import processing.IChanges;
+import processing.utility.*;
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java:4]: No need of `processing.IChanges`. Please confirm.

File: src/processing/CommunicateChange.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		for (Position pos : modifiedPos) {
+			Position curPos = new Position(pos);
+			Intensity intensity;
+			
+			// gets the objectId at pos
+			ObjectId objectId = ClientBoardState
+					.maps
+					.getMostProbableObjectId(
+							new ArrayList <Position> (Arrays.asList(pos))
+							);
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java]: Doubtful regarding multi-line conventions. Please confirm.

File: src/processing/CommunicateChange.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+				
+				if (pixels == null || pixels.size() <= 0) {
+					// Ideally, it should not happen
+					// error should be logged
+					intensity = new Intensity(255, 255, 255);
+				}
+				else {
+					// An Object has same color on all of its pixels
+					// gets the intensity of the object from the first pixel
+					intensity = new Intensity(pixels.get(0).intensity);
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java]: multiline comments in /* * */

File: src/processing/CommunicateChange.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		// gets the handler for the subscriber
+		IChanges handler = ClientBoardState.identifierTohandler.get(identifier);
+		// sends the subscriber the modified pixels
+		handler.getChanges(modifiedPixels);
+	}
+	
+	// UI will subscribe for any changes 
+	public static void subscribeForChanges(Identifier identifier, IChanges handler) {
+		ClientBoardState.identifierTohandler.put(identifier, handler);
+	}
```
Comment (RathoreDevansh08) [inline src/processing/CommunicateChange.java:98]: Is `ClentBoardState.identifierTohandler` defined uptill now...I cannot find it. Also ideally according to conventions, it should be `identifierToHandler`. Please just check once.

**Review by RathoreDevansh08 — CHANGES_REQUESTED** [review]
edits marked. May have further corrections.

File: src/processing/CommunicateChange.java (hunk 1 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java:5]: Added

File: src/processing/CommunicateChange.java (hunk 2 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java:31]: Done

File: src/processing/CommunicateChange.java (hunk 3 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java:36]: There is no convention regarding this. So, it's fine.

File: src/processing/CommunicateChange.java (hunk 4 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java:4]: Yes, no need.

File: src/processing/CommunicateChange.java (hunk 5 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java]: It's fine

File: src/processing/CommunicateChange.java (hunk 6 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java]: Done

File: src/processing/CommunicateChange.java (hunk 7 above)
Comment (iamrakesh28) [reply src/processing/CommunicateChange.java:98]: Made a private member for that

**Comment by iamrakesh28** [comment]
Can be reviewed and merged

**Review by RathoreDevansh08 — APPROVED** [review]

**Pull request merged**